/***************************************************************************
 *
 * Copyright (c) 2015 Baidu, Inc. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 **************************************************************************/
// Author: Wen Xiang <wenxiang@baidu.com>
//

package baidu.flume;

message PbTransferEncodingField {
    enum Type {
        RECORD = 1;
        GLOBAL_KEY = 2;
        LOCAL_PARTITION = 3;
        FIXED_LENGTH_KEY = 4;
        VARIABLE_LENGTH_KEY = 5;
        TAILING_KEY = 6;
    }

    required string identity = 1;  // executor/dispatcher identity, mainly for debug
    required Type type = 2;
    optional uint32 length = 3;  // set for FIXED_LENGTH_KEY
}

message PbTransferEncoder {
    enum TagEncoding {
        UNIQUE = 1;
        UINT8 = 2;
        UINT16 = 3;
    }

    required PbTransferEncodingField field = 1;
    required uint32 tag = 2;
    required TagEncoding tag_encoding = 3;
    optional PbTransferEncoder successor = 4;
}

message PbTransferDecoder {
    required PbTransferEncodingField field = 1;
    repeated PbTransferDecoder successor = 2;
}
